lib/Mojo/mysql/Connection.pm

```diff
@@ -0,0 +1,797 @@
+package Mojo::mysql::Connection;
+use Mojo::Base 'Mojo::EventEmitter';
+
+use utf8;
+use Encode qw(_utf8_off _utf8_on);
+use Digest::SHA qw(sha1);
+use Scalar::Util 'weaken';
+use Mojo::IOLoop;
+
+has host => 'localhost';
+has port => 3306,
+has username => '';
+has password => '';
+has database => '';
+
+has options => sub { {
+  found_rows => 0, multi_statements => 0, utf8 => 1,
+  connect_timeout => 10, query_timeout => 0 } };
+
+has _state => 'disconnected';
+
+use constant DEBUG => $ENV{MOJO_MYSQL_DEBUG} // 0;
+
+use constant {
+  CLIENT_CAPABILITY => [ qw(
+    LONG_PASSWORD FOUND_ROWS LONG_FLAG CONNECT_WITH_DB
+    NO_SCHEMA COMPRESS ODBC LOCAL_FILES
+    IGNORE_SPACE PROTOCOL_41 INTERACTIVE SSL
+    IGNORE_SIGPIPE TRANSACTIONS RESERVED SECURE_CONNECTION
+    MULTI_STATEMENTS MULTI_RESULTS PS_MULTI_RESULTS PLUGIN_AUTH
+    CONNECT_ATTRS PLUGIN_AUTH_LENENC_CLIENT_DATA CAN_HANDLE_EXPIRED_PASSWORDS SESSION_TRACK
+    DEPRECATE_EOF) ],
+
+  SERVER_STATUS => [ qw(
+    STATUS_IN_TRANS STATUS_AUTOCOMMIT RESERVED MORE_RESULTS_EXISTS
+    STATUS_NO_GOOD_INDEX_USED STATUS_NO_INDEX_USED STATUS_CURSOR_EXISTS STATUS_LAST_ROW_SENT
+    STATUS_DB_DROPPED STATUS_NO_BACKSLASH_ESCAPES STATUS_METADATA_CHANGED QUERY_WAS_SLOW
+    PS_OUT_PARAMS STATUS_IN_TRANS_READONLY SESSION_STATE_CHANGED) ],
+
+  FIELD_FLAG => [ qw(
+    NOT_NULL PRI_KEY UNIQUE_KEY MULTIPLE_KEY
+    BLOB UNSIGNED ZEROFILL BINARY
+    ENUM AUTO_INCREMENT TIMESTAMP SET) ],
+
+  CHARSET => {
+    UTF8 => 33, BINARY => 63, ASCII => 65 },
+
+  DATATYPE => {
+    DECIMAL => 0x00, TINY => 0x01, SHORT => 0x02, LONG => 0x03,
+    FLOAT => 0x04, DOUBLE => 0x05,
+    NULL => 0x06, TIMESTAMP => 0x07,
+    LONGLONG => 0x08, INT24 => 0x09,
+    DATE => 0x0a, TIME => 0x0b, DATETIME => 0x0c, YEAR => 0x0d, NEWDATE => 0x0e,
+    VARCHAR => 0x0f, BIT => 0x10,
+    NEWDECIMAL => 0xf6, ENUM => 0xf7, SET => 0xf8,
+    TINY_BLOB => 0xf9, MEDIUM_BLOB => 0xfa, LONG_BLOB => 0xfb, BLOB => 0xfc,
+    VAR_STRING => 0xfd, STRING => 0xfe, GEOMETRY => 0xff },
+};
+
+use constant {
+  REV_CHARSET => { reverse %{CHARSET()} },
+  REV_DATATYPE => { map { chr(DATATYPE->{$_}) => $_ } keys %{DATATYPE()} },
+};
+
+use constant SEQ => {
+  connect => {
+    connected => '_recv_handshake',
+    handshake => '_send_auth',
+    auth => '_recv_ok',
+  },
+  query => {
+    idle => '_send_query',
+    query => '_recv_query_responce',
+    field => '_recv_field',
+    result => '_recv_row',
+  },
+  ping => {
+    idle => '_send_ping',
+    ping => '_recv_ok',
+  },
+  disconnect => {
+    idle => '_send_quit',
+    quit => '_recv_ok'
+  }
+};
+
+
+sub _flag_list($$;$) {
```

Comment:
Two things: It seems like a lot of these methods are "protocol" functions and not tied to $self. Could you just do ```_flag_list(...)``` instead of calling them as method?

The other things is that I really don't like prototypes. ($$;$) I want a really good reason to keep them around.